### PR TITLE
[api][SIMS #1453]Moved notifications to shared library.

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
- Moved notifications services to the shared library under `@sims/services/notifications`.
- Created the module and marked it as `@Global` importing it only on `app.module` on API and removed it from others.
- Included existing file upload notifications in the DB transaction.
- Added option to send the student restrictions notifications right away or later (delayed option).

### CODE REVIEW

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 1]
```diff
@@ -24,88 +25,105 @@ export class GCNotifyActionsService {
    * This method is used to send email notification to SABC
    * when a student uploads documents and submits it in the file uploader screen.
    * @param notification input parameters to generate the notification.
-   * @param userId id of the user who will receive the message.
-   * @param auditUserId id of the user creating the notification.
+   * @param auditUserId user that should be considered the one that is causing the changes.
+   * @param entityManager optional repository that can be provided, for instance,
+   * to execute the command as part of an existing transaction. If not provided
+   * the local repository will be used instead.
```
Comment (ann-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:31]: we can remove the return comment

Comment (andrewsignori-aot) [reply sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:31]: Removed.

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 2]
```diff
@@ -129,12 +145,29 @@ export class GCNotifyActionsService {
...
       notificationsToSend,
       auditUserId,
-      entityManager,
+      options?.entityManager,
     );
 
-    // TODO: Temporary code to be removed once queue/schedulers are in place.
-    for (const id of notificationsIds) {
-      await this.notificationService.sendEmailNotification(id);
+    if (!options?.notificationsDelayed) {
```
Comment (ann-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:151]: what happens when `notificationsDelayed` is true

Comment (andrewsignori-aot) [reply sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:151]: The message is saved to the DB only. The expectation is that there will be a job running to take all the ones that were never sent and try to send them again.

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 3]
```diff
@@ -129,12 +145,29 @@ export class GCNotifyActionsService {
...
+      // Execute all the notifications in parallel and return the promisees
+      // to allow the method to await them all.
+      const notificationPromisees = notificationsIds.map((notificationId) =>
```
Comment (andrepestana-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:154]: "notificationPromises" ?

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 4]
```diff
@@ -129,12 +145,29 @@ export class GCNotifyActionsService {
...
+        // Not intended to be used for large volume of notifications. If a large
+        // volume is expected, the notificationsDelayed must be set to true
+        this.notificationService.sendEmailNotification(
+          notificationId,
+          options?.entityManager,
+        ),
+      );
+
+      try {
+        // Wait all promisees.
```
Comment (andrepestana-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:164]: "wait for all promises."

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 5]
```diff
@@ -129,12 +145,29 @@ export class GCNotifyActionsService {
...
 
-    // TODO: Temporary code to be removed once queue/schedulers are in place.
-    for (const id of notificationsIds) {
-      await this.notificationService.sendEmailNotification(id);
+    if (!options?.notificationsDelayed) {
+      // Execute all the notifications in parallel and return the promisees
+      // to allow the method to await them all.
+      const notificationPromisees = notificationsIds.map((notificationId) =>
+        // Not intended to be used for large volume of notifications. If a large
+        // volume is expected, the notificationsDelayed must be set to true
```
Comment (andrepestana-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:156]: I don't mind about it but as it has been pointed out in one of my other PRs, I noticed the comment is missing the period punctuation mark here. 😄

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 6]
```diff
@@ -24,88 +25,103 @@ export class GCNotifyActionsService {
...
       },
     };
+
     // Save notification into notification table.
-    const notificationSaved = await this.notificationService.saveNotification(
-      userId,
-      NotificationMessageType.StudentFileUpload,
-      payload,
+    const [notificationId] = await this.notificationService.saveNotifications(
+      [notificationToSend],
```
Comment (guru-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:60]: 👍

**Review by andrepestana-aot — COMMENTED** [review]
Good word!

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts (hunk 5 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts]: Yes, it is valid to point it out. To add some context why it is needed, in the past we started to try to add more comments and the sentences were hard to follow sometimes because there was no upper case, commas, or punctuation at a level that we need to read (or at least I need to read) many times the comments to try to "decode" the idea, which almost invalidates the comment sometimes.
The general idea is simple, try to follow regular sentences in which the period is part of it.
So, are the periods really important, not that much, but they are in the middle of a paragraph.
To keep it easy, a general rule, it must look like an English sentence 😉 

This particular one was addressed already.

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts]: Changed.

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts]: Changed.

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @andrewsignori-aot

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts (hunk 4 above)
Comment (andrepestana-aot) [reply sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts]: in this case we need the "for"... in "wait for something" 
https://dictionary.cambridge.org/grammar/british-grammar/wait-or-wait-for

**Review by ann-aot — APPROVED** [review]
👍 Good work @andrewsignori-aot

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts]: The reason why I added the comment on https://github.com/bcgov/SIMS/pull/1565#discussion_r1029856948 was to provide a general understanding of the reason why we do it to allow us to ponderate how far we should go.
From the moment that we start to talk more about the grammar than the code itself is when I guess that we start to lose it a bit 😉

Comment (andrepestana-aot) [reply sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts]: For me it's fine. Don't worry. I would not comment about it if you were not changing that line.

**Review by andrepestana-aot — APPROVED** [review]
Thanks for the changes!
